src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs

```diff
@@ -0,0 +1,431 @@
+            }
+
+            private string[]? _ids;
+            private int[]? _byteOffsets;
+            private int[]? _lengths;
+            private string _tzFileDir;
+            private string _tzFilePath;
+
+            private static string GetApexTimeDataRoot()
+            {
+                string? ret = Environment.GetEnvironmentVariable("ANDROID_TZDATA_ROOT");
+                if (!string.IsNullOrEmpty(ret))
+                {
+                    return ret;
+                }
+
+                return "/apex/com.android.tzdata";
+            }
+
+            private static string GetApexRuntimeRoot()
+            {
+                string? ret = Environment.GetEnvironmentVariable("ANDROID_RUNTIME_ROOT");
+                if (!string.IsNullOrEmpty(ret))
+                {
+                    return ret;
+                }
+
+                return "/apex/com.android.runtime";
+            }
+
+            public AndroidTzData()
+            {
+
+                string[] tzFileDirList = new string[] {GetApexTimeDataRoot() + "/etc/tz/", // Android 10+, TimeData module where the updates land
+                                                       GetApexRuntimeRoot() + "/etc/tz/", // Android 10+, Fallback location if the above isn't found or corrupted
+                                                       Environment.GetEnvironmentVariable("ANDROID_DATA") + "/misc/zoneinfo/",
+                                                       Environment.GetEnvironmentVariable("ANDROID_ROOT") + DefaultTimeZoneDirectory};
+                foreach (var tzFileDir in tzFileDirList)
+                {
+                    string tzFilePath = Path.Combine(tzFileDir, TimeZoneFileName);
+                    if (LoadData(tzFilePath))
+                    {
+                        _tzFileDir = tzFileDir;
+                        _tzFilePath = tzFilePath;
+                        return;
+                    }
+                }
+
+                throw new TimeZoneNotFoundException(SR.TimeZoneNotFound_ValidTimeZoneFileMissing);
+            }
+
+            private bool LoadData(string path)
+            {
+                if (!File.Exists(path))
+                {
+                    return false;
+                }
+                try {
+                    using (FileStream fs = File.OpenRead(path))
+                    {
+                        LoadTzFile(fs);
+                    }
+                    return true;
+                }
+                catch (IOException) {}
+
+                return false;
+            }
+
+            private unsafe void LoadTzFile(Stream fs)
+            {
+                int headerSize = 24;
+                Span<byte> buffer = stackalloc byte[headerSize];
+                int bytesRead = 0;
+                int bytesLeft = headerSize;
+
+                while (bytesLeft > 0)
+                {
+                    int b = fs.Read(buffer);
+                    if (b == 0)
```

Comment:
This should probably throw if we reach the end of the file prematurely.

Reply:
Thanks! Throwing `throw new InvalidOperationException(SR.Format(SR.InvalidOperation_ReadTZError, _tzFilePath, 0, HeaderSize, bytesRead, HeaderSize));` similar to `LoadEntryAt`'s exception thrown.